### DISCUSSION TITLE
Switches, Reversible pistons and Generals

### DISCUSSION THREAD

**Discussion by Jean-Gabriel1982** [GrandOrgue/GoOdf#70] [General]
Question stupid but I don't understand the use of those options .. 

Of course, I don't want to say that it's useless, it's not my point, 😃  but my question is what is it for? How to use it and for what effect?

Thanks !!

**Comment by larspalo** [comment]
Switches are useful when you need to do something that's more complex than what's possible to do with only one single stop. Thus it's a button that allows control of (possibly) multiple other elements. One such example is adding drawstop noises. But many other complex things (like the french Appel d'Anches combination systems, or "Sperrventil") can be done with switches. If you don't know a use case for a switch - then you most likely shouldn't use one either because it's most likely un-necessary.

Reversible pistons are a relic of old. Nowadays a similar functionality can be accomplished with just another GUI Element (styled differently) on the panel.

Generals are the old style odf defined general combinations. Basically, the only need for them nowadays is when you have to model fixed combinations or do some kind of "tricks" that involve changing the state of many other elements in possibly different ways. They can be protected so that the user cannot modify them. But "generally" the built in versions of the generals that GO provide, and you can create as GUI Elements (of setter type) are to be preferred.

**Comment by Psi58** [comment]
Lars' answer is perfect - clear and with sufficient detail/context to aid understanding. But how will a new user in a couple of months, with a similar question, easily find it? 

GO is a fantastic project but there's a steep learning curve for new users, especially if they want to adapt, or create new, organs. There are 2 things that would really help.

(1) A glossary of terminology <ins>in terms that a non-expert new user would understand</ins> (sorry, the Help file is too brief, tends to assume some deeper knowledge, and doesn't contain examples). This discussion illustrates the need for that. There's no way that the distinction between stops, switches, ranks etc can be easily understood intuitively. And at the moment we (I...) have to search through help files, Github discussions, various forums, Facebook etc to try to find answers - a single 'official' glossary which could be contributed to by all would be great. And, if it existed, Lars could have just highlighted the link...

(2) A repository of ODF code samples. I'm learning by reviewing existing samplesets and scouring all the sources mentioned above. But it's very hit and miss  - trying to find the answer to a specific issue (eg [my latest bugbear](https://github.com/GrandOrgue/grandorgue/discussions/1690)) seems harder than it should be. If there was a library of code samples that everyone was encouraged to donate to, the knowledge and capability of the whole community would improve, which has to be beneficial for the project.

Sadly I don't have the knowledge or capability to take the lead in creating either of these, but I would certainly be an active supporter/contributor if they existed.

**Reply by larspalo** [reply]
> GO is a fantastic project but there's a steep learning curve for new users, especially if they want to adapt, or create new, organs.

I cannot but agree! It has taken me somewhere around 15 years to get to my current state in regards to GrandOrgue. And I'm still learning! But I'm also trying to work on improving the situation for others as best I can.

I'm currently trying to, in various ways, contribute to or maintain GrandOrgue, GOODF and LoopAuditioneer (and occasionally other projects) as well as working on already recorded sample sets to get them released and plan for recording new organs. Also I'm working professionally as a church musician and have a family with five children to take care of. As you might imagine other activities also tend to take time - hunting, (hobby) farming, general property maintenance around the house etc - just to name a few. Thus the time simply doesn't exist to do everything I'd like to do and some prioritizing is necessary.

Improvement suggestions of the help, especially in the form of pull requests, would certainly be welcome. It's really not that difficult to use the features GitHub provides. It could also be possible to add wiki pages, but improved help that comes bundled with the software is likely best.

I've actually contemplated creating a sample set with demonstrations of various possibilities, but maybe a repo of its own under GrandOrgue would be a better option, where users both can submit examples and learn from others.

I can warmly recommend watching some of my screencasts at https://www.youtube.com/playlist?list=PLNz9AJhjubTdw0pV2nZC_XfphJhJV5ly3 to learn more about using GOODF for creating .organ files for GrandOrgue. If someone has suggestions for topics to cover I'd be happy to give it a try.

**Reply by Psi58** [reply]
Well that was a major omission on my part! OF COURSE the screencasts - both GOODF and GO - are invaluable for understanding how things work, how to use the tools etc. In fact I would say that watching them convinced me to have a go in the first place despite my very rudimentary programming knowledge. I periodically watch them again to remind myself of all the goodies available. What, IMHO, they're not ideal for is answering a specific question. I'm never going to remember that at 8 minutes 23 second into a screencast a particular topic starts, for example. But maybe a document that indexes all the topics covered and where to find them would help...?

Anyway I'd just like to say that I'm hugely appreciative of all the material that DOES exist, and my ideas were certainly not a 'Let's create more work for Lars' suggestion. The huge advantage of open source ought to be that the work can be shared, and hopefully there are others who have the skills to take ownership of different aspects of the project. I will see if I can work out how to use pull requests, as suggested.
